bigtable-emulator-maven-plugin doesn't work offline

When running without an internet connection, using the Maven emulator plugin fails with
```
[ERROR] Failed to execute goal com.google.cloud.bigtable:bigtable-emulator-maven-plugin:1.0.0:start (default) on project ***: Execution default of goal com.google.cloud.bigtable:bigtable-emulator-maven-plugin:1.0.0:start failed: Failed to query gcloud sdk components. Is gcloud sdk installed?: Failed to get components list: ERROR: (gcloud.components.list) Failed to fetch component listing from server. Check your network settings and try again. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.google.cloud.bigtable:bigtable-emulator-maven-plugin:1.0.0:start (default) on project ***: Execution default of goal com.google.cloud.bigtable:bigtable-emulator-maven-plugin:1.0.0:start failed: Failed to query gcloud sdk components. Is gcloud sdk installed?
```
because `gcloud components list` needs an internet connection.

Can the online requirement be relaxed, either by skipping this call (and just look for the emulator in the expected path), or ignoring this specific error if an internet connection is not available.

--- a/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java
+++ b/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java
@@ -207,25 +207,37 @@ public class BigtableOptions implements Serializable {
         return;
       }
 
-      String[] hostPort = emulatorHost.split(":");
+      enableEmulator(emulatorHost);
+    }
+
+    public Builder enableEmulator(String emulatorHostAndPort) {
+      String[] hostPort = emulatorHostAndPort.split(":");
       Preconditions.checkArgument(hostPort.length == 2,
           "Malformed " + BIGTABLE_EMULATOR_HOST_ENV_VAR + " environment variable: " +
-          emulatorHost + ". Expecting host:port.");
+          emulatorHostAndPort + ". Expecting host:port.");
 
       int port;
       try {
         port = Integer.parseInt(hostPort[1]);
       } catch (NumberFormatException e) {
         throw new RuntimeException("Invalid port in " + BIGTABLE_EMULATOR_HOST_ENV_VAR +
-            " environment variable: " + emulatorHost);
+            " environment variable: " + emulatorHostAndPort);
       }
+      enableEmulator(hostPort[0], port);
+      return this;
+    }
+
+    public Builder enableEmulator(String host, int port) {
+      Preconditions.checkArgument(host != null && !host.isEmpty(), "Host cannot be null or empty");
+      Preconditions.checkArgument(port > 0, "Port must be positive");
       setUsePlaintextNegotiation(true);
       setCredentialOptions(CredentialOptions.nullCredential());
-      setDataHost(hostPort[0]);
-      setAdminHost(hostPort[0]);
+      setDataHost(host);
+      setAdminHost(host);
       setPort(port);
 
-      LOG.info("Connecting to the Bigtable emulator at " + emulatorHost);
+      LOG.info("Connecting to the Bigtable emulator at " + host + ":" + port);
+      return this;
     }
 
     public BigtableOptions build() {

--- a/bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/BigtableEnv.java
+++ b/bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/BigtableEnv.java
@@ -47,6 +47,7 @@ class BigtableEnv extends SharedTestEnv {
       BigtableOptionsFactory.BIGTABLE_PORT_KEY,
       BigtableOptionsFactory.BIGTABLE_HOST_KEY,
       BigtableOptionsFactory.BIGTABLE_ADMIN_HOST_KEY,
+      BigtableOptionsFactory.BIGTABLE_EMULATOR_HOST_KEY,
       BigtableOptionsFactory.PROJECT_ID_KEY,
       BigtableOptionsFactory.INSTANCE_ID_KEY,
       BigtableOptionsFactory.BIGTABLE_USE_BULK_API,

--- a/bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/test_env/BigtableEnv.java
+++ b/bigtable-hbase-parent/bigtable-hbase-integration-tests/src/test/java/com/google/cloud/bigtable/hbase/test_env/BigtableEnv.java
@@ -41,6 +41,7 @@ class BigtableEnv extends SharedTestEnv {
       BigtableOptionsFactory.BIGTABLE_PORT_KEY,
       BigtableOptionsFactory.BIGTABLE_HOST_KEY,
       BigtableOptionsFactory.BIGTABLE_ADMIN_HOST_KEY,
+      BigtableOptionsFactory.BIGTABLE_EMULATOR_HOST_KEY,
       BigtableOptionsFactory.PROJECT_ID_KEY,
       BigtableOptionsFactory.INSTANCE_ID_KEY,
       BigtableOptionsFactory.BIGTABLE_USE_BULK_API,

--- a/bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java
+++ b/bigtable-hbase-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java
@@ -64,6 +64,12 @@ public class BigtableOptionsFactory {
   /** Constant <code>BIGTABLE_HOST_KEY="google.bigtable.endpoint.host"</code> */
   public static final String BIGTABLE_HOST_KEY = "google.bigtable.endpoint.host";
 
+  /**
+   * Constant <code>BIGTABLE_HOST_KEY="google.bigtable.emulator.endpoint.host"</code>. Values should
+   * be of format: `host:port`
+   */
+  public static final String BIGTABLE_EMULATOR_HOST_KEY = "google.bigtable.emulator.endpoint.host";
+
   /** Constant <code>PROJECT_ID_KEY="google.bigtable.project.id"</code> */
   public static final String PROJECT_ID_KEY = "google.bigtable.project.id";
   /** Constant <code>INSTANCE_ID_KEY="google.bigtable.instance.id"</code> */
@@ -294,6 +300,11 @@ public class BigtableOptionsFactory {
     bigtableOptionsBuilder.setPort(port);
     bigtableOptionsBuilder.setUsePlaintextNegotiation(
       configuration.getBoolean(BIGTABLE_USE_PLAINTEXT_NEGOTIATION, false));
+
+    String emulatorHost = configuration.get(BIGTABLE_EMULATOR_HOST_KEY);
+    if (emulatorHost != null) {
+      bigtableOptionsBuilder.enableEmulator(emulatorHost);
+    }
 
     setBulkOptions(configuration, bigtableOptionsBuilder);
     setChannelOptions(configuration, bigtableOptionsBuilder);

